lib/java/com/google/android/material/drawable/DrawableUtils.java

```diff
@@ -190,63 +194,124 @@ private static Drawable createTintableMutatedDrawableIfNeeded(
-      // If the top layer is smaller than the bottom layer in both its width and height, keep top
-      // layer's size.
-      topLayerNewWidth = topLayerDrawable.getIntrinsicWidth();
-      topLayerNewHeight = topLayerDrawable.getIntrinsicHeight();
+
+    // Centers the top layer.
+    if (VERSION.SDK_INT >= VERSION_CODES.M) {
+      drawable.setLayerGravity(1, Gravity.CENTER);
     } else {
-      float topLayerRatio =
-          (float) topLayerDrawable.getIntrinsicWidth() / topLayerDrawable.getIntrinsicHeight();
-      float bottomLayerRatio =
-          (float) bottomLayerDrawable.getIntrinsicWidth()
-              / bottomLayerDrawable.getIntrinsicHeight();
+      // On Pre-M there's no layer gravity support, so we use layer insets to adjust the top layer
+      // position.
+      final int topLayerIntrinsicWidth = topLayerDrawable.getIntrinsicWidth();
+      final int topLayerIntrinsicHeight = topLayerDrawable.getIntrinsicHeight();
+      final int bottomLayerIntrinsicWidth = bottomLayerDrawable.getIntrinsicWidth();
+      final int bottomLayerIntrinsicHeight = bottomLayerDrawable.getIntrinsicHeight();
+
+      final int horizontalInset = max((bottomLayerIntrinsicWidth - topLayerIntrinsicWidth) / 2, 0);
+      final int verticalInset = max((bottomLayerIntrinsicHeight - topLayerIntrinsicHeight) / 2, 0);
+      drawable.setLayerInset(1,
+          horizontalInset, verticalInset, horizontalInset, verticalInset);
+    }
+
+    return drawable;
+  }
+
+  /**
+   * Composites two drawables, returning a drawable instance of {@link LayerDrawable},
+   * with the top layer centered and scaled with the aspect ratio preserved to the size of
+   * the bottom layer.
+   * <p>
+   * If any of the drawables is null, this method will return the other.
+   *
+   * @param bottomLayerDrawable the drawable to be on the bottom layer
+   * @param topLayerDrawable the drawable to be on the top layer
+   * @param topLayerDesiredWidth top layer desired width in pixels, or -1 to use the intrinsic
+   *     width if possible.
+   * @param topLayerDesiredHeight top layer desired height in pixels, or -1 to use the intrinsic
+   *     height if possible.
+   */
+  @RequiresApi(VERSION_CODES.M)
+  @Nullable
+  public static Drawable compositeDrawables(
+      @Nullable Drawable bottomLayerDrawable, @Nullable Drawable topLayerDrawable,
+      @Px int topLayerDesiredWidth, @Px int topLayerDesiredHeight) {
+
+    if (bottomLayerDrawable == null) {
+      return topLayerDrawable;
+    }
+
+    if (topLayerDrawable == null) {
+      return bottomLayerDrawable;
+    }
+
+    final int topLayerIntrinsicWidth = topLayerDrawable.getIntrinsicWidth();
+    final int topLayerIntrinsicHeight = topLayerDrawable.getIntrinsicHeight();
+    final int bottomLayerIntrinsicWidth = bottomLayerDrawable.getIntrinsicWidth();
+    final int bottomLayerIntrinsicHeight = bottomLayerDrawable.getIntrinsicHeight();
+
+    int topLayerNewWidth =
```

Comment:
Declare int constants instead of using magic number -1.

Reply:
Android doesn't seem to have `Drawable.INTRINSIC_SIZE` or something...

Are you proposing to create something like that?
```java
@RestrictTo(Scope.LIBRARY_GROUP)
public final class DrawableUtils {

  public static final int INTRINSIC_SIZE = -1;

  ...
}
```

Reply:
Yes, that's what I mean. And maybe call it UNSPECIFIED_SIZE?

Reply:
Well, technically, "intrinsic size" cannot be called unspecified. And I think `use INTRINSIC_SIZE to use the intrinsic size` a bit clearer than `use UNSPECIFIED_SIZE to use the intrinsic size`. But if you insist, I can change it.